drivers/led/lp5569.c

```diff
@@ -0,0 +1,169 @@
+/*
+ * Copyright (c) 2023 Phytec Messtechnik GmbH.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT ti_lp5569
+
+/**
+ * @file
+ * @brief LP5569 LED controller
+ *
+ * The LP5569 is a 9-channel LED driver that communicates over I2C.
+ */
+
+#include <zephyr/drivers/i2c.h>
+#include <zephyr/drivers/led.h>
+#include <zephyr/device.h>
+#include <zephyr/kernel.h>
+
+#define LOG_LEVEL CONFIG_LED_LOG_LEVEL
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(lp5569);
+
+#include "led_context.h"
+
+#define LP5569_NUM_LEDS			9
+#define LP5569_DEFAULT_LED_CURRENT	0xAF	/* 17.5 mA (Reset Default) */
+#define LP5569_DEFAULT_DUTY_CYCLE_ON	0xFF	/* Duty cycle in percent */
+
+/* General Registers */
+#define LP5569_CONFIG			0x00
+#define LP5569_CHIP_EN			(1 << 6)
+
+#define LP5569_MISC			0x2F
+#define LP5569_POWERSAVE_EN		(1 << 5)
+
+/* PWM Registers for controlling the duty-cycle */
+#define LP5569_LED0_PWM			0x16
+#define LP5569_LED1_PWM			0x17
+#define LP5569_LED2_PWM			0x18
+#define LP5569_LED3_PWM			0x19
+#define LP5569_LED4_PWM			0x1A
+#define LP5569_LED5_PWM			0x1B
+#define LP5569_LED6_PWM			0x1C
+#define LP5569_LED7_PWM			0x1D
+#define LP5569_LED8_PWM			0x1E
+
+/* Current Registers */
+#define LP5569_LED0_CURRENT		0x22
+#define LP5569_LED1_CURRENT		0x23
+#define LP5569_LED2_CURRENT		0x24
+#define LP5569_LED3_CURRENT		0x25
+#define LP5569_LED4_CURRENT		0x26
+#define LP5569_LED5_CURRENT		0x27
+#define LP5569_LED6_CURRENT		0x28
+#define LP5569_LED7_CURRENT		0x29
+#define LP5569_LED8_CURRENT		0x2A
+
+struct lp5569_config {
+	struct i2c_dt_spec bus;
+	const struct device *vin_supply;
+};
+
+struct lp5569_data {
+	struct led_data dev_data;
+};
+
+static int lp5569_led_set_brightness(const struct device *dev, uint32_t led,
+				     uint8_t brightness)
+{
+	const struct lp5569_config *config = dev->config;
+	uint8_t val;
+
+	if (led >= LP5569_NUM_LEDS || brightness > 100) {
+		return -EINVAL;
+	}
+
+	/* Map 0-100 % to 0-255 pwm register value */
+	val = brightness * 2.55;
+
+	if (i2c_reg_write_byte_dt(&config->bus, LP5569_LED0_PWM + led, val)) {
+		LOG_ERR("LED reg update failed");
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static inline int lp5569_led_on(const struct device *dev, uint32_t led)
+{
+	const struct lp5569_config *config = dev->config;
+
```

Comment:
Eventually you should make it a call to `lp5569_led_set_brightness`. This would avoid some code duplication.